### PR TITLE
Make ShutdownTest.LanguageServerShutsDownIfParentDies stable

### PR DESCRIPTION
**Pull request by keyboardDrummer**
Resolve the test instability that failed this run: https://github.com/dafny-lang/dafny/runs/7071695232?check_suite_focus=true

<small>By submitting this pull request, I confirm that my contribution is made under the terms of the [MIT license](https://github.com/dafny-lang/dafny/blob/master/LICENSE.txt).</small>

### CODE REVIEW

File: Source/DafnyLanguageServer.Test/Various/ShutdownTest.cs [hunk 1]
```diff
@@ -71,12 +71,21 @@ public class ShutdownTest {
...
-        languageServer.Kill();
-      }, "Language server should have killed itself if the parent is gone.");
+      for (int attempts = 0; attempts < 10; attempts++) {
+        // Give the language server time to kill itself
+        await Task.Delay(100);
+
+        try {
+          Process.GetProcessById(int.Parse(languageServerProcessId));
+        } catch {
+          return;
```
Comment (fabiomadge) [inline Source/DafnyLanguageServer.Test/Various/ShutdownTest.cs:82]: Shouldn't this be a `continue`?

File: Source/DafnyLanguageServer.Test/Various/ShutdownTest.cs [hunk 2]
```diff
@@ -71,12 +71,21 @@ public class ShutdownTest {
...
-      Assert.ThrowsException<ArgumentException>(() => {
-        var languageServer = Process.GetProcessById(int.Parse(languageServerProcessId));
-        languageServer.Kill();
-      }, "Language server should have killed itself if the parent is gone.");
+      for (int attempts = 0; attempts < 10; attempts++) {
+        // Give the language server time to kill itself
+        await Task.Delay(100);
+
+        try {
+          Process.GetProcessById(int.Parse(languageServerProcessId));
```
Comment (fabiomadge) [inline Source/DafnyLanguageServer.Test/Various/ShutdownTest.cs:80]: And a `break` here?

File: Source/DafnyLanguageServer.Test/Various/ShutdownTest.cs [hunk 3]
```diff
@@ -71,12 +71,21 @@ public class ShutdownTest {
...
+        } catch {
+          return;
+        }
+      }
+
+      var languageServer = Process.GetProcessById(int.Parse(languageServerProcessId));
```
Comment (fabiomadge) [inline Source/DafnyLanguageServer.Test/Various/ShutdownTest.cs:86]: Couldn't we do this assignment in the loop?

**Review by fabiomadge — COMMENTED** [review]
Why is that preferable to `Task.Dalay(1000)`?

File: Source/DafnyLanguageServer.Test/Various/ShutdownTest.cs (hunk 2 above)
Comment (keyboardDrummer) [reply Source/DafnyLanguageServer.Test/Various/ShutdownTest.cs:80]: If this try doesn't throw an exception then the current iteration has failed, so we need to keep trying.

File: Source/DafnyLanguageServer.Test/Various/ShutdownTest.cs (hunk 1 above)
Comment (keyboardDrummer) [reply Source/DafnyLanguageServer.Test/Various/ShutdownTest.cs:82]: The catch is the happy case, in which the language server has killed itself and the test should pass.

File: Source/DafnyLanguageServer.Test/Various/ShutdownTest.cs (hunk 3 above)
Comment (keyboardDrummer) [reply Source/DafnyLanguageServer.Test/Various/ShutdownTest.cs:86]: This is the cleanup code once we've decided the language server won't kill itself so we're going to manually do it.

**Comment by keyboardDrummer** [comment]
> Why is that preferable to `Task.Delay(1000)`?

Because that takes a long time.

File: Source/DafnyLanguageServer.Test/Various/ShutdownTest.cs (hunk 3 above)
Comment (fabiomadge) [reply Source/DafnyLanguageServer.Test/Various/ShutdownTest.cs:86]: Sure, but why the assignment.

**Comment by fabiomadge** [comment]
> Because that takes a long time. The philosophy is, if you're waiting for an event that has no event listener, then keep looping while you wait for it until your test times out, instead of waiting an arbitrary time, since the latter produce unstable tests.

But why is 10*100 more stable than 1000? I'm not convinced this added complexity buys us much.

File: Source/DafnyLanguageServer.Test/Various/ShutdownTest.cs (hunk 3 above)
Comment (keyboardDrummer) [reply Source/DafnyLanguageServer.Test/Various/ShutdownTest.cs:86]: The value is used on the next line. I'm not sure I understand the question.

Comment (fabiomadge) [reply Source/DafnyLanguageServer.Test/Various/ShutdownTest.cs:86]: Yes, but you also generate it in L80. Might as well use that.

Comment (keyboardDrummer) [reply Source/DafnyLanguageServer.Test/Various/ShutdownTest.cs:86]: Okay. I just looked into doing that but it creates warnings around assignment that I'd rather not override.

**Comment by keyboardDrummer** [comment]
> Yes, but this is not what this PR is proposing. 

Changed it to wait 10ms for 500 times.

> I don't think a second is too much of a problem.

I think a second is too much, because we'll have thousands of unit tests. Having them take a second when they only take 100ms on average seems wasteful.

**Comment by fabiomadge** [comment]
> I think a second is too much, because we'll have thousands of unit tests. Having them take a second when they only take 100ms on average seems wasteful.

Sure, but we're only discussing a single test. Maybe we could come up with some abstraction, if that is a pattern we expect to deploy often. At least in case of the unit tests, we have some parallelism which would alleviate some of the problems of excessive waiting.

**Review by fabiomadge — APPROVED** [review]

File: Source/DafnyLanguageServer.Test/Various/ShutdownTest.cs (hunk 3 above)
Comment (fabiomadge) [reply Source/DafnyLanguageServer.Test/Various/ShutdownTest.cs:86]: Thanks for looking into it.

**Pull request merged**